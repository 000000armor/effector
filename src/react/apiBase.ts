import {
  Store,
  is,
  createNode,
  step,
  clearNode,
  scopeBind,
  Scope,
  Unit,
  Event,
  Node,
} from 'effector'
import React from 'react'
import {useSyncExternalStore} from 'use-sync-external-store/shim'
import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/shim/with-selector'
import {throwError} from './throw'
import {createWatch} from './createWatch'
import {withDisplayName} from './withDisplayName'

const stateReader = <T>(store: Store<T>, scope?: Scope) =>
  scope ? scope.getState(store) : store.getState()
const basicUpdateFilter = <T>(upd: T, oldValue: T) => upd !== oldValue
const keysEqual = (a?: readonly any[], b?: readonly any[]) => {
  if (!a || !b || a.length !== b.length) return false

  let isEqual = true

  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      isEqual = false
      break
    }
  }

  return isEqual
}

export function useStoreBase<State>(store: Store<State>, scope?: Scope) {
  if (!is.store(store)) throwError('expect useStore argument to be a store')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )
  const currentValue = useSyncExternalStore(subscribe, read, read)

  return currentValue
}

export function useUnitBase<Shape extends {[key: string]: Unit<any>}>(
  shape: Shape,
  scope?: Scope,
) {
  const isShape = !is.unit(shape) && typeof shape === 'object'
  const normShape = isShape ? shape : {unit: shape}
  const isList = Array.isArray(normShape)
  const entries = Object.entries(normShape)
  const [current, storeMap, stores] = React.useMemo(() => {
    const initial: Record<string, any> = isList ? [] : {}
    const stores: Store<any>[] = []
    const storeMap: Record<string, string> = {}
    for (const [key, value] of entries) {
      if (is.store(value)) {
        if (stores.includes(value)) {
          throwError(`useUnit store at key "${key}" is already exists in shape`)
        }
        stores.push(value)
        /** note that this allows only one occurence of the store */
        storeMap[(value as any).graphite.id] = key
        initial[key] = stateReader(value, scope)
      } else {
        if (!is.unit(value)) throwError('expect useUnit argument to be a unit')
        initial[key] = scope ? scopeBind(value as Event<any>, {scope}) : value
      }
    }
    return [{ref: initial}, storeMap, stores]
  }, [...entries.flat(), scope])
  const subscribe = React.useCallback(
    (cb: () => void) => {
      const seq = [
        step.compute({
          fn(value, _, stack) {
            const storeId = stack.parent!.node.id
            const storeKey = storeMap[storeId]
            current.ref = isList
              ? [...(current.ref as any[])]
              : {...current.ref}
            current.ref[storeKey] = value
          },
        }),
        /**
         * 'effect' priority cannot be batched
         * so we wait other effects to finish and then run batching
         * this is not an essential to work
         * just a fine tuning of a really low priority (ui) update
         *
         * also note that step.run({fn}) is an alias for
         * step.compute({fn, priority: 'effect'})
         * and since 22 version fn might be ommited if not used
         * */
        step.compute({priority: 'effect'}),
        step.compute({priority: 'sampler', batch: true}),
        step.run({fn: () => cb()}),
      ]
      if (scope) {
        const node = createNode({node: seq})
        const scopeLinks: {[_: string]: Node[]} = (scope as any).additionalLinks
        const storeLinks: Node[][] = []
        stores.forEach(store => {
          const id = (store as any).graphite.id
          const links = scopeLinks[id] || []
          scopeLinks[id] = links
          links.push(node)
          storeLinks.push(links)
        })
        return () => {
          storeLinks.forEach(links => {
            const idx = links.indexOf(node)
            if (idx !== -1) links.splice(idx, 1)
          })
          clearNode(node)
        }
      } else {
        const node = createNode({
          node: seq,
          parent: stores,
          family: {owners: stores},
        })
        return () => {
          clearNode(node)
        }
      }
    },
    [current],
  )
  const read = React.useCallback(
    () => (isShape ? current.ref : current.ref.unit),
    [current],
  )
  return useSyncExternalStore(subscribe, read, read)
}

export function useStoreMapBase<State, Result, Keys extends ReadonlyArray<any>>(
  [configOrStore, separateFn]: [
    configOrStore:
      | {
          store: Store<State>
          keys: Keys
          fn(state: State, keys: Keys): Result
          updateFilter?: (update: Result, current: Result) => boolean
        }
      | Store<State>,
    separateFn?: (state: State, keys: Keys) => Result,
  ],
  scope?: Scope,
): Result {
  let fn: (state: State, keys: Keys) => Result
  let updateFilter: (update: Result, current: Result) => boolean =
    basicUpdateFilter
  let store: Store<State>
  let keys: Keys
  if (separateFn) {
    fn = separateFn
    store = configOrStore as Store<State>
    keys = [] as unknown as Keys
  } else {
    fn = (configOrStore as any).fn
    store = (configOrStore as any).store
    keys = (configOrStore as any).keys
    updateFilter = (configOrStore as any).updateFilter || basicUpdateFilter
  }
  if (!is.store(store)) throwError('useStoreMap expects a store')
  if (!Array.isArray(keys)) throwError('useStoreMap expects an array as keys')
  if (typeof fn !== 'function') throwError('useStoreMap expects a function')

  const subscribe = React.useCallback(
    (cb: () => void) => createWatch(store, cb, scope),
    [store, scope],
  )
  const read = React.useCallback(
    () => stateReader(store, scope),
    [store, scope],
  )

  const stateRef = React.useRef<State>()
  const valueRef = React.useRef<Result>()
  const keysRef = React.useRef(keys)

  const value = useSyncExternalStoreWithSelector(
    subscribe,
    read,
    read,
    state => {
      if (stateRef.current !== state || !keysEqual(keysRef.current, keys)) {
        const result = fn(state, keys)

        stateRef.current = state
        keysRef.current = keys

        // skip update, if undefined
        // just like original store or previous implementation
        if (result !== undefined) {
          valueRef.current = result
        }
      }

      return valueRef.current as Result
    },
    (current, update) => !updateFilter(update, current),
  )

  return value
}
export function useListBase<T>(
  list: Store<T[]>,
  renderItem:
    | {
        keys?: any[]
        fn(item: T, index: number): React.ReactNode
        getKey?: (item: T) => string
      }
    | ((item: T, index: number) => React.ReactNode),
  scope?: Scope,
): React.ReactNode {
  let keys = [] as any[]
  let fn
  let getKey: (item: T) => string
  if (typeof renderItem === 'object' && renderItem !== null) {
    if (renderItem.keys) keys = renderItem.keys
    fn = renderItem.fn
    if (renderItem.getKey) getKey = renderItem.getKey
  } else {
    fn = renderItem
  }
  if (!is.store(list)) throwError('expect useList first argument to be a store')
  if (typeof fn !== 'function')
    throwError("expect useList's renderItem to be a function")
  if (!Array.isArray(keys)) throwError("expect useList's keys to be an array")
  const Item = React.useMemo(() => {
    const Item = withDisplayName(
      `${list.shortName || 'Unknown'}.Item`,
      (
        props:
          | {index: number; keys: any[]; keyVal: never; value: never}
          | {index: never; keys: any[]; keyVal: string; value: T},
      ) => {
        const {index, keys, keyVal, value} = props
        const isKeyed = !!fnRef.current[1]
        if (isKeyed) {
          return fnRef.current[0](value, keyVal as any)
        }
        const item = useStoreMapBase(
          [
            {
              store: list,
              keys: [index, ...keys],
              fn: (list, keys) => list[keys[0]],
            },
          ],
          scope,
        )
        return fnRef.current[0](item, index)
      },
    )
    return React.memo(Item)
  }, [list, scope, !!getKey!])
  const fnRef = React.useRef([fn, getKey!] as const)
  fnRef.current = [fn, getKey!]
  const keysSelfMemo = React.useMemo(() => keys, keys)
  if (getKey!) {
    return useStoreBase(list, scope).map(value => {
      const key = fnRef.current[1](value)
      return React.createElement(Item, {
        keyVal: key,
        key,
        keys: keysSelfMemo,
        value,
      })
    })
  } else {
    const length = useStoreMapBase(
      [
        {
          store: list,
          keys: [list],
          fn: list => list.length,
        },
      ],
      scope,
    )
    return Array.from({length}, (_, i) =>
      React.createElement(Item, {
        index: i,
        key: i,
        keys: keysSelfMemo,
      }),
    )
  }
}
