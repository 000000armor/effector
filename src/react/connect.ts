import React from 'react'

import {Store} from 'effector'
import {createReactState} from './createReactState'

export const connect = <State>(Component: React.ComponentType<any>) => (
  store: Store<State>,
) => createReactState(store, Component)
