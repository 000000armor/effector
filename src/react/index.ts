export {createStoreConsumer} from './createStoreConsumer'
export {createComponent} from './createComponent'
export {connect} from './connect'
export {createGate, useGate} from './createGate'
export {useEvent, useList, useStore, useStoreMap, useUnit} from './nossr'
