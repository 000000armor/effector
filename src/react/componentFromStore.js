//@flow

import * as React from 'react'
import {
  createEvent,
  createStore,
  type Event,
  type Store,
  type Subscription,
} from 'effector'

export function componentFromStore<Props: {...}>(
  propsToVdom: (props: Store<Props>) => Store<React.Node>,
) {
  class ComponentFromStore extends React.Component<Props, {vdom: React.Node, ...}> {
    state /*: {vdom: React.Node, ...}*/ = {vdom: null}

    propsEmitter: Event<Props> = createEvent()

    // Stream of props
    props$: Store<Props> = createStore((null: any))
      .on(this.propsEmitter, (_, props) => props)
      .map(props => {
        if (props) return props
        /*::return props*/
      })

    // Stream of vdom
    vdom$ = propsToVdom(this.props$)

    subscription: Subscription

    componentWillMount() {
      // Subscribe to child prop changes so we know when to re-render
      this.subscription = this.vdom$.watch(vdom => {
        this.setState({vdom})
      })
      this.propsEmitter(this.props)
    }

    componentWillReceiveProps(nextProps: Props) {
      // Receive new props from the owner
      this.propsEmitter(nextProps)
    }

    shouldComponentUpdate(
      nextProps: Props,
      nextState /*: {vdom: React.Node, ...}*/,
    ) {
      return nextState.vdom !== this.state.vdom
    }

    componentWillUnmount() {
      this.props$.off(this.propsEmitter)
      this.subscription.unsubscribe()
    }

    render() {
      return this.state.vdom
    }
  }
  return ComponentFromStore
}
