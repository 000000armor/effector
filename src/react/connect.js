//@flow

import * as React from 'react'

import type {Store} from 'effector'
import {createReactState} from './createReactState'

export function connect<State: Object, Com: React.ComponentType<*>>(
  Component: Com,
): (
  store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
  return store => createReactState(store, Component)
}
