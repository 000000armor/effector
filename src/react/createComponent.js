//@flow

import * as React from 'react'
import {
  type Store,
  isStore,
  //$todo add invariant to effector public typings
  invariant,
  createStoreObject,
  createEvent,
} from 'effector'
import type {StoreView} from './index.h'

type Unsubscribe = () => void

export function createComponent<Props: {}, State>(
  shape: Store<State> | {+[key: string]: Store<any> | any},
  renderProp: (props: Props, state: State) => React.Node,
): StoreView<State, Props> {
  let store: Store<any>
  if (isStore(shape)) {
    store = shape
  } else {
    invariant(
      typeof shape === 'object' && shape !== null,
      'shape should be store or object with stores',
    )
    store = createStoreObject(shape)
  }
  const mounted = createEvent(`${store.shortName}.View mounted`)
  const unmounted = createEvent(`${store.shortName}.View unmounted`)
  return class RenderComponent extends React.Component<
    Props,
    {currentState: State},
  > {
    static displayName = `${store.shortName}.View`
    static mounted = mounted
    static unmounted = unmounted
    _unsubscribe: Unsubscribe | null = null
    _hasUnmounted: boolean = false
    state = {currentState: store.getState()}
    componentDidMount() {
      mounted({props: this.props, state: store.getState()})
      this.subscribe()
    }
    componentWillUnmount() {
      unmounted({props: this.props, state: store.getState()})
      this.unsubscribe()
      this._hasUnmounted = true
    }
    render() {
      return renderProp(this.props, this.state.currentState)
    }
    subscribe() {
      this._unsubscribe = store.watch((state: State) => {
        if (this._hasUnmounted) {
          return
        }

        this.setState(prevState => {
          if (state === prevState.currentState) {
            return null
          }
          return {currentState: state}
        })
      })
    }
    unsubscribe() {
      if (typeof this._unsubscribe === 'function') {
        this._unsubscribe()
      }

      this._unsubscribe = null
    }
  }
}
