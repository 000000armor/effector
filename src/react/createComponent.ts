import * as React from 'react'
import {Store, is, combine, createEvent} from 'effector'
import {useStore} from './useStore'
import {useIsomorphicLayoutEffect} from './useIsomorphicLayoutEffect'
import {StoreView} from './index.h'
import {withDisplayName} from './withDisplayName'
import {throwError} from './throw'

export function createComponent<Props, State>(
  shape:
    | Store<State>
    | {[key: string]: Store<any> | any}
    | ((props: Props) => Store<State>),
  renderProp: (props: Props, state: State) => React.ReactNode,
): StoreView<State, Props> {
  let storeFn: (props: Props) => Store<any>
  let store: Store<any>
  if (is.store(shape)) {
    store = shape
  } else if (typeof shape === 'function') {
    storeFn = shape as any
    console.error('storeFactory is deprecated')
  } else {
    if (typeof shape === 'object' && shape !== null) {
      store = combine(shape)
    } else throwError('shape should be a store or object with stores')
  }
  let storeName = 'Unknown'
  //@ts-ignore
  if (store && store.shortName) {
    storeName = store.shortName
  }
  const mounted = createEvent<any>(`${storeName}.View mounted`)
  const unmounted = createEvent<any>(`${storeName}.View unmounted`)

  //prettier-ignore
  const instanceFabric: (props: Props) => Store<any> =
    typeof shape === 'function'
      //@ts-ignore
      ? storeFn
      : () => store
  function RenderComponent(props: Props) {
    const propsRef = React.useRef(props)
    const ownStore = React.useMemo(() => instanceFabric(props), [])
    const state = useStore(ownStore)
    useIsomorphicLayoutEffect(() => {
      mounted({props: propsRef.current, state: ownStore.getState()})
      return () => {
        unmounted({props: propsRef.current, state: ownStore.getState()})
      }
    }, [])
    const result = renderProp(props, state)
    propsRef.current = props
    return result
  }
  RenderComponent.mounted = mounted
  RenderComponent.unmounted = unmounted
  return withDisplayName(`${storeName}.View`, RenderComponent)
}
