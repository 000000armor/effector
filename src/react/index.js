//@flow

import * as React from 'react'

import type {Store} from 'effector'

export function createReactState<State>(
 store: Store<State>,
): {
 Provider: React.ComponentType<{children: React.Node}>,
 Consumer: React.ComponentType<{
  children: (current: {
   value: State,
   ref: Store<State>,
  }) => React.Node,
 }>,
} {
 const {Provider, Consumer} = React.createContext({
  value: store.getState(),
  ref: store,
 })

 class StoreProvider extends React.Component<
  {
   children: React.Node,
  },
  {
   current: {
    value: State,
    ref: Store<State>,
   },
  },
 > {
  state = {
   current: {
    value: store.getState(),
    ref: store,
   },
  }
  componentDidMount() {
   store.watch(value => this.setState({current: {value, ref: store}}))
  }
  render() {
   return <Provider value={this.state.current}>{this.props.children}</Provider>
  }
 }

 return {
  Provider: StoreProvider,
  Consumer,
 }
}

export function createStoreComponent<
 State: Object,
 Com: React.ComponentType<*>,
>(
 store: Store<State>,
 Component: Com,
): React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 const {Provider, Consumer} = createReactState(store)
 const ConnectedComponent = props => (
  <Provider>
   <Consumer>{({value}) => <Component {...props} {...value} />}</Consumer>
  </Provider>
 )
 ConnectedComponent.displayName = `Connect(${Component.displayName
  || 'Unknown'})`
 return ConnectedComponent
}

export function connect<State: Object, Com: React.ComponentType<*>>(
 Component: Com,
): (
 store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 return store => createStoreComponent(store, Component)
}
