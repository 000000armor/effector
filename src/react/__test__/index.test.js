//@flow

import * as React from 'react'
import {shallow, mount, render} from 'enzyme'
import {
 createEvent,
 createStore,
 type Event,
 type Store,
} from '../../effector'
import {createReactState} from '..'

test('basic', () => {
 const store = createStore('hello')
 const changeText = createEvent('change text')

 const {Provider, Consumer} = createReactState(store)

 const tree = mount(
  <Provider>
   <Consumer>
    {({value, ref}) => <span>{value}</span>}
   </Consumer>
  </Provider>
 )
 expect(tree.text()).toMatchSnapshot()
 tree.unmount()
})
