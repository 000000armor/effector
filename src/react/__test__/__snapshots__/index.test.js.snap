// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`click counter 1`] = `"Counter: 0Increment"`;

exports[`click counter 2`] = `"Counter: 1Increment"`;

exports[`refined api 1`] = `"Text: fooCounter: 1"`;

exports[`refined api 2`] = `"Text: barCounter: 1"`;
